### PR TITLE
[Auto Benchmarks] Add stat collection

### LINKED ISSUE #2345: [Auto Benchmarks] Add stat collection
### What is this task and why do we need to work on it?

After the new run metrics/stats are defined in #2343, we need to implement them. We need these to enable more detailed benchmarking runs. 

### What work will need to be done to complete this task?

This involves going through the code and adding counters for the data we want to look at. It does not involve publishing anything to the orchestrator, as that will be solved later.

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

The new decided metrics all have values that can be printed at the end of a benchmarking run.

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by dailinsubjam**
Closes #2345 

### This PR: 
- Tracks a list of stats for benchmarking. For more details:  #2345  .
- Adds arguments  to parameterize benchmarking.
- Adds scripts for benchmarking runs.
- Fixes `readme` for examples.

### This PR does not: 
- Involve publishing anything to the orchestrator, as that will be solved later.
- Include neat output for benchmarking results, it's in the format of error messages now, later we might move to CSV file.
- Track `commit` that benchmarks running, will be added once integrated to CI.
- Track `notes field`, also to be investigated after CI integration.
- Carefully design the parameter we want for each run. The script includes roughly designed runs based on my local results. Also local machines only support limited threads, number of nodes cannot be too large (20 is fine, 50 is not). 
- Consider any VID-related factor.

### Key places to review: 
`crates/examples/infra/mod.rs`

### How to test this PR:
`./scripts/benchmarks.sh`

### Local results example:
With `just async_std example all-webserver -- --config_file ./crates/orchestrator/run-config.toml --total_nodes 20 --da_committee_size 5 --transactions_per_round 10 --transaction_size 512 --rounds 100` I'll get:
```
Failed views: 0, Total views: 100, num_successful_commits: 100
100 rounds completed in 44.871671042s
Avg latency: 15 sec, Minimum latency: 1 sec, Maximum latency: 45 sec, Throughput: 292067 bytes/sec
```

### CODE REVIEW

File: crates/examples/infra/mod.rs
Comment (jparr721) [inline crates/examples/infra/mod.rs:973]: In the changeset above I see you added a comment to explain the `+8`. I presume that its use here is for the same purpose, however, since it's >300 lines below the initial comment, do you think it'd be worth adding a comment here? Or perhaps making a named constant like (`TIMESTAMP_BYTES_SIZE`)?

**Review by jparr721 — COMMENTED** [review]
Howdy! I'm the _new guy_. Just had one thing that I was curious about. Feel free to ignore it though if it isn't relevant.

File: crates/examples/infra/mod.rs
Comment (dailinsubjam) [reply crates/examples/infra/mod.rs:973]: Welcome to the team! I should add a comment here, the reason is that `transaction_size` in the config means appending this size of bytes to a default transaction with 8 bytes (see `create_random_transaction`).

**Review by bfish713 — COMMENTED** [review]
This looks good to me, would be good if someone a bit more familiar with the examples/orchestrator could take a look

File: scripts/benchmarks.sh
Comment (dailinsubjam) [reply scripts/benchmarks.sh:1]: Yes for certain! We have a script to run locally now, but I'll need to coordinate with integration team to see how scripts should be like to run in AWS.

File: crates/examples/combined/orchestrator.rs
Comment (elliedavidson) [inline crates/examples/combined/orchestrator.rs:28]: Can we make it so we set the orchestrator url to something we pass in as arguments?  We may need to change the port sometimes when we run on AWS.  We can have it default to localhost:4444 if no arguments are provided though, since having a default will make it more ergonomic.  This is the only change request I have for the PR.

File: crates/examples/infra/mod.rs
Comment (elliedavidson) [inline crates/examples/infra/mod.rs:90]: Love this function!

File: crates/examples/infra/mod.rs
Comment (elliedavidson) [inline crates/examples/infra/mod.rs:589]: Do we actually ever emit these events?  I don't think we do?  I think they are left over from a long time ago before the event-based model.  (Though that isn't related to this PR)

File: crates/examples/infra/mod.rs
Comment (elliedavidson) [inline crates/examples/infra/mod.rs:609]: Thank you for the "in seconds" comment!

File: crates/orchestrator/README.md
Comment (elliedavidson) [inline crates/orchestrator/README.md:5]: Thank you for updating the docs!

File: scripts/benchmarks.sh
Comment (elliedavidson) [inline scripts/benchmarks.sh:1]: Love this!

**Review by elliedavidson — CHANGES_REQUESTED** [review]
I'm only requesting one change - that we allow the orchestrator to run on a parametrizable port.  All the other comments are optional changes or just comments.

**Review by jparr721 — COMMENTED** [review]
A few questions. Feel free to ignore whatever isn't helpful.

File: crates/example-types/src/block_types.rs
Comment (elliedavidson) [reply crates/example-types/src/block_types.rs:173]: No, we should really standardize it :) We're open to suggestions - personally I like including the "get" prefix.

Comment (dailinsubjam) [reply crates/example-types/src/block_types.rs:173]: Thanks for pointing this out!

Comment (dailinsubjam) [reply crates/example-types/src/block_types.rs:173]: This part is mostly for `BlockPayload`, I only added a single function, so I tend to not change it too much (which is out of scope of this PR). But I'll clean up the name for posting / getting things from orchestrator according to this principle.

File: crates/examples/infra/mod.rs
Comment (dailinsubjam) [reply crates/examples/infra/mod.rs:164]: Oh this is for default setting, I'd better remove this.

Comment (dailinsubjam) [reply crates/examples/infra/mod.rs]: BTW, they're all very helpful questions!

File: crates/examples/infra/mod.rs
Comment (dailinsubjam) [reply crates/examples/infra/mod.rs:623]: Yeah I've deleted one EventType, but after taking a close look I also deleted `NextLeaderViewTimeout` since it's never emitted, for others I just keep them. This is also the reason I added `EventType::ViewTimeout` to track failed views, together with `assert!(total_num_views <= (failed_num_views + num_successful_commits + 2))`.

**Review by jparr721 — APPROVED** [review]
I'll defer to @elliedavidson before you merge this, but this looks good to me!

**Review by bfish713 — APPROVED** [review]
I've been using this locally and having no issues.  I think it's good to go!

**Review by elliedavidson — APPROVED** [review]

**Pull request merged**
